tests/hvm_cpu.rs

```diff
@@ -0,0 +1,384 @@
+use hvm_lang::term::{parser, DefNames, DefinitionBook, Term};
+use hvmc::{ast::*, run::NumericOp, *};
+use insta::{assert_debug_snapshot, assert_snapshot};
+use std::fs;
+
+fn load_file(file: &str) -> String {
+  let path = format!("{}/tests/programs/{}", env!("CARGO_MANIFEST_DIR"), file);
+  fs::read_to_string(path).unwrap()
+}
+
+// Parses code and generate Book from hvm-core syntax
+fn parse_core(code: &str) -> Book {
+  do_parse_book(code)
+}
+
+// Parses code and generate DefinitionBook from hvm-lang syntax
+fn parse_lang(code: &str) -> DefinitionBook {
+  parser::parse_definition_book(code).unwrap()
+}
+
+// Loads file and generate DefinitionBook from hvm-lang syntax
+fn load_lang(file: &str) -> DefinitionBook {
+  let code = load_file(file);
+  parse_lang(&code)
+}
+
+// For every pair in the map, replaces all matches of a string the other string
+fn replace_template(mut code: String, map: &[(&str, &str)]) -> String {
+  for (from, to) in map {
+    code = code.replace(from, to);
+  }
+  code
+}
+
+trait Normal {
+  type Result;
+  fn normal(self, size: usize) -> Self::Result;
+}
+
+impl Normal for Book {
+  type Result = (run::Net, Net);
+
+  fn normal(self, size: usize) -> Self::Result {
+    let book = book_to_runtime(&self);
+    let mut rnet = run::Net::new(size);
+    rnet.boot(name_to_val("main"));
+    rnet.normal(&book);
+
+    let net = net_from_runtime(&rnet);
+    (rnet, net)
+  }
+}
+
+impl Normal for DefinitionBook {
+  type Result = (Term, DefNames, hvm_lang::RunInfo);
+
+  fn normal(self, size: usize) -> Self::Result {
+    hvm_lang::run_book(self, size).unwrap()
+  }
+}
+
+#[test]
+fn test_era_era() {
+  let net = parse_core("@main = * & * ~ *");
+  let (rnet, net) = net.normal(16);
+  assert_snapshot!(show_net(&net), @"*");
+  assert_debug_snapshot!(rnet.rewrites(), @"2");
+}
+
+#[test]
+fn test_con_dup() {
+  let net = parse_core("@main = root & (x x) ~ {2 * root}");
+  let (rnet, net) = net.normal(16);
+  assert_snapshot!(show_net(&net), @"(b b)");
+  assert_debug_snapshot!(rnet.rewrites(), @"5");
+}
+
+#[test]
+fn test_bool_and() {
+  let book = parse_core(
+    "
+    @true = (b (c b))
+    @fals = (b (c c))
+    @and  = ((b (@fals c)) (b c))
+    @main = root & @and ~ (@true (@fals root))
+  ",
+  );
+  let (rnet, net) = book.normal(64);
+
+  assert_snapshot!(show_net(&net), @"(b (c c))");
+  assert_debug_snapshot!(rnet.rewrites(), @"8");
+}
+
+#[test]
+fn test_church_mul() {
+  let book = load_lang("church_mul.hvm");
+  let (term, defs, info) = book.normal(64);
+
+  assert_snapshot!(show_net(&info.net), @"({2 ({2 b c} d) {3 (d e) (e {2 c f})}} (b f))");
+  assert_snapshot!(term.to_string(&defs), @"λa λb (a (a (a (a (a (a b))))))");
+}
+
+#[test]
+fn test_neg_fusion() {
+  let book = load_lang("neg_fusion.hvm");
+  let (_, _, info) = book.normal(512);
+
+  assert_snapshot!(show_net(&info.net), @"(b (* b))");
+  assert_debug_snapshot!(info.stats.rewrites.total_rewrites(), @"153");
```

Comment:
I think you can do this
```suggestion
  let (_, _, Info { net }) = book.normal(512);

  assert_snapshot!(show_net(&net), @"(b (* b))");
  assert_debug_snapshot!(net.rewrites(), @"153");
```

Reply:
isn't it a good idea to `assert_snapshot` of the hvm-lang term as well?

Reply:
The net inside the info is a different type than the net with the rewrites method

Reply:
Will add the assert_snapshot of the hvm-lang term